core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -1250,6 +1362,147 @@ public boolean move() {
         public boolean cannotRemain() {
             return isDecisionTaken() && canRemainDecision.type() == Type.NO;
         }
+
+        /**
+         * Gets the individual node-level decisions that went into making the final decision as represented by
+         * {@link #getFinalDecisionType()}.  The map that is returned has the node id as the key and a {@link WeightedDecision}.
+         */
+        @Nullable
+        public Map<String, WeightedDecision> getNodeDecisions() {
+            return nodeDecisions;
+        }
+    }
+
+    /**
+     * Represents a decision to move a started shard to form a more optimally balanced cluster.
+     */
+    public static final class RebalanceDecision extends RelocationDecision {
+        /** a constant representing no decision taken */
+        public static final RebalanceDecision NOT_TAKEN = new RebalanceDecision(null, null, null, null, null, Float.POSITIVE_INFINITY);
+
+        @Nullable
+        private final Decision canRebalanceDecision;
+        @Nullable
+        private final Map<String, NodeRebalanceDecision> nodeDecisions;
+        private float currentWeight;
+
+        protected RebalanceDecision(Decision canRebalanceDecision, Type finalDecision, String finalExplanation) {
+            this(canRebalanceDecision, finalDecision, finalExplanation, null, null, Float.POSITIVE_INFINITY);
+        }
+
+        protected RebalanceDecision(Decision canRebalanceDecision, Type finalDecision, String finalExplanation,
+                                    String assignedNodeId, Map<String, NodeRebalanceDecision> nodeDecisions, float currentWeight) {
+            super(finalDecision, finalExplanation, assignedNodeId);
+            this.canRebalanceDecision = canRebalanceDecision;
+            this.nodeDecisions = nodeDecisions != null ? Collections.unmodifiableMap(nodeDecisions) : null;
```

Comment:
Do you think we should keep `null` as the identifier here, or can we use an empty map instead? It'd be nice if we could so we'd avoid NPEs in the future; what do you think?